runtime/tests/tests.rs

```diff
@@ -222,35 +280,441 @@ fn generic_asset_transfer_works_with_fee_exchange() {
+			// and then trap.
+			let contract_call = Call::Contracts(pallet_contracts::Call::call(
+				contract_address.clone(), // newly created contract address
+				0,                        // transfer value in
+				5_000_000_000,            // gas limit
+				vec![],
+			));
+			let fee_exchange = FeeExchange::V1(FeeExchangeV1 {
+				asset_id: CENNZ_ASSET_ID,
+				max_payment: bob_max_funds,
+			});
+			let contract_call_extrinsic = sign(CheckedExtrinsic {
+				signed: Some((bob(), signed_extra(0, 0, None, Some(fee_exchange)))),
+				function: contract_call,
+			});
+
+			// This only shows transaction fee payment success, not gas payment success
+			assert!(Executive::apply_extrinsic(contract_call_extrinsic).is_ok());
+
+			let block_events = frame_system::Module::<Runtime>::events();
+			let events = vec![
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(0),
+					event: Event::pallet_contracts(RawEvent::CodeStored(code_hash.into())),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(0),
+					event: Event::system(frame_system::Event::ExtrinsicSuccess(DispatchInfo {
+						weight: 10000,
+						class: DispatchClass::Normal,
+						pays_fee: true,
+					})),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(1),
+					event: Event::pallet_contracts(RawEvent::Transfer(alice(), contract_address.clone(), 0)),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(1),
+					event: Event::pallet_contracts(RawEvent::Instantiated(alice(), contract_address.clone())),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(1),
+					event: Event::system(frame_system::Event::ExtrinsicSuccess(DispatchInfo {
+						weight: 10000,
+						class: DispatchClass::Normal,
+						pays_fee: true,
+					})),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(2),
+					event: Event::crml_cennzx_spot(crml_cennzx_spot::RawEvent::AssetPurchase(
+						CENNZ_ASSET_ID,
+						CENTRAPAY_ASSET_ID,
+						bob(),
+						2_587_750_030_067,
+						2_580_010_000_000,
+					)),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(2),
+					event: Event::crml_cennzx_spot(crml_cennzx_spot::RawEvent::AssetPurchase(
+						CENNZ_ASSET_ID,
+						CENTRAPAY_ASSET_ID,
+						bob(),
+						421_261_139,
+						420_001_135,
+					)),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(2),
+					event: Event::pallet_contracts(RawEvent::Dispatched(contract_address.clone(), false)),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::ApplyExtrinsic(2),
+					event: Event::system(frame_system::Event::ExtrinsicSuccess(DispatchInfo {
+						weight: 10000,
+						class: DispatchClass::Normal,
+						pays_fee: true,
+					})),
+					topics: vec![],
+				},
+			];
+			assert_eq!(block_events[..7], events[..7]);
```

Comment:
It's not clear to me how these sets of events are showing the GA transfer has failed (assuming comments at the top of the test are still correct).
Suggest adding some comments to explain the assert logic